### ISSUE TITLE
Access or Save all Models in Objective

### ISSUE THREAD

**Issue by rm17tink** [optuna/optuna#3446]
### Environment

Environment
Optuna version: 2.10.0
Python version: 3.6
OS: Windows 10
(Optional) Other libraries and their versions: lightgbm = 3.3.2

### Description

Forgive me if this is novice programming. But if there a way I can store and access each trial's information including but not limited to the train_auc as shown below. I am also interested in storing each trials parameters, number of columns, AUC's and graph them afterwards for an audience. Any item I can access, keep or record as the trials run would be helpful for presentation purposes. 

Thanks 

### Error messages, stack traces, or logs

```shell
No error, but maybe user error.
```

### Steps to reproduce

```python

def objective_no_cv(trial, X_train, y_train, X_valid, y_valid):
    global model
    
    X_train = X_train
    y_train = y_train
    X_valid = X_valid
    y_valid = y_valid
    
    param_grid = {
        #         "device_type": trial.suggest_categorical("device_type", ['gpu']),
        "n_estimators": trial.suggest_categorical("n_estimators", [10000]),
        "verbose": -1,  # mutes the [WARNING] output
        "boosting_type": "gbdt", #dart
         "seed": 42,
        #"scale_pos_weight": 583, #(makes performance terrible)
        #"is_unbalance": "true",  #added ( makes performance terrible)
        "learning_rate": trial.suggest_float("learning_rate", 0.1, 0.3),
        "num_leaves": trial.suggest_int("num_leaves", 20, 3000, step=20),
        "max_depth": trial.suggest_int("max_depth", 3, 12),
        "min_data_in_leaf": trial.suggest_int("min_data_in_leaf", 200, 10000, step=100),
        "max_bin": trial.suggest_int("max_bin", 200, 300),
        "lambda_l1": trial.suggest_int("lambda_l1", 0, 100, step=5),
        "lambda_l2": trial.suggest_int("lambda_l2", 0, 100, step=5),
        "min_gain_to_split": trial.suggest_float("min_gain_to_split", 0, 15),
        "bagging_fraction": trial.suggest_float(
            "bagging_fraction", 0.2, 0.95, step=0.1
        ),
        "bagging_freq": trial.suggest_categorical("bagging_freq", [1]),
        "feature_fraction": trial.suggest_float(
            "feature_fraction", 0.2, 0.95, step=0.1
        ),
    }
    model = lgb.LGBMClassifier(objective="binary", **param_grid)
    model.fit(
            X_train,
            y_train,
            eval_set=[(X_valid, y_valid)],
            eval_metric="auc",
            early_stopping_rounds=100,
            callbacks=[
                LightGBMPruningCallback(trial, "auc")
            ],  # Add a pruning callback
        )
        
    train_preds = model.predict_proba(X_train)
    valid_preds = model.predict_proba(X_valid)
    
    auc_scores_train = roc_auc_score(y_valid, train_preds[:,1])
    auc_scores_valid = roc_auc_score(y_valid, valid_preds[:,1])
    
    return auc_scores_valid

# 
# allows you to save the best model and store it in best_booster(allegedly)
# gives you the best model to use again
def callback(study, trial):
    global best_booster
    if study.best_trial == trial:
        best_booster = model
   
```

**Comment by nzw0301** [comment]
I'm not sure I could understand the issue description, but let me leave my comments.

I suppose we would like to store additional metrics and models that are used later for analysis or something like that. 
If such metric or object is pickable, we can store it as a trial attribute by using [`set_trial_attr`](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.trial.Trial.html#optuna.trial.Trial.set_user_attr):

```python
def objective_no_cv(trial, X_train, y_train, X_valid, y_valid):
     ...
     # store auc store on the training dataset.
     trial.set_trial_attr("auc_scores_train", auc_scores_train) 
```

After optimization, we access the stored value by calling `FrozenTrial.user_attr`.

For non-pickable cases, I think the following way is straightforward:
https://optuna.readthedocs.io/en/stable/faq.html#how-to-save-machine-learning-models-trained-in-objective-functions

**Comment by rm17tink** [comment]
I dont mean to open this up again... but is there a way for me to choose a trial and repurpose the info the same way i call study.best_

I ran the dataframe and prefer trial 1 instead of the best trial. Id like to replace all the info with trial 1 and save that as the best. Is it possibe so that I can then access the below code the same way? 

```

print(f"\tBest Value (auc): {study.best_value:.5f}")
print(f"\tBest Trial Number: {study.best_trial.number}")
print(f"\tNumber of features in Model: {best_booster.n_features_in_} ") 
print(f"\tBest params:")
for key, value in study.best_params.items():
    print(f"\t\t{key}: {value}")
```
